lib/Support/GzipStream.cpp

```diff
@@ -0,0 +1,99 @@
+/// \file GzipStream.cpp
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "revng/Support/Assert.h"
+#include "revng/Support/GzipStream.h"
+
+#include "zlib.h"
+
+constexpr size_t OutputBufferSize = 4 * 1024;
+static_assert(OutputBufferSize <= UINT_MAX);
+
+constexpr int windowBits = 15 /* 2**15 bytes (32k) of window */
+                           + 16 /* Magic offset for gzip */;
+
+template<int (*Next)(z_streamp, int)>
+static void zlibCopyStream(z_stream &Stream,
+                           llvm::raw_ostream &OS,
+                           const llvm::ArrayRef<uint8_t> Buffer) {
+  llvm::SmallVector<uint8_t> OutBuffer;
+  OutBuffer.resize_for_overwrite(OutputBufferSize);
+  const char *OutBufferPtr = reinterpret_cast<const char *>(OutBuffer.data());
+
+  size_t RemainingInput = Buffer.size();
+
+  Stream.next_in = const_cast<uint8_t *>(Buffer.data());
```

Comment:
`const_cast` are absolutely illegal, please check everywhere.
Use `MutableArrayRef`.

Reply:
That's not the intention, it's the `libz` api's fault that it only accepts `byte *` but **does not change the contents**.

Reply:
OK, then add a comment about this.

We should put `const_cast` in check conventions and use `NONEXTLINELINT` in such cases.

Reply:
Added comment, noted adding the rule to check-conventions in TODO